src/hepattn/models/loss.py

```diff
@@ -194,36 +215,53 @@ def mask_focal_cost(pred_logits, targets, gamma=2.0, input_pad_mask=None, sample
-        object_valid_mask: [batch_size, num_objects] - mask indicating valid target objects
-        input_pad_mask: [batch_size, num_inputs] - mask indicating valid inputs
-        sample_weight: Optional sample weights for each element.  Recommended to use focal instead.
+        pred_logits (torch.Tensor): Float logits of shape `[batch_size, num_objects, num_constituents]`.
+        targets (torch.Tensor): Binary targets in `{0,1}` or floats in `[0,1]` of shape `[batch_size, num_objects, num_constituents]`.
+        object_valid_mask (torch.Tensor, optional): Boolean mask of shape `[batch_size, num_objects]`
+            indicating which objects are valid for each sample in the batch. Defaults to all ones.
+        input_pad_mask (torch.Tensor, optional): Boolean or float mask of shape `[batch_size, num_constituents]`
+            indicating valid inputs (e.g., non-padded constituents) per batch sample. Defaults to all ones.
+        sample_weight (torch.Tensor, optional): Optional weighting tensor broadcastable to `[batch_size, num_objects, num_constituents]`
+            applied elementwise after BCE computation.
 
     Returns:
-        loss: Scalar tensor representing the binary cross-entropy loss
+        torch.Tensor: Scalar loss averaged over valid objects and inputs in the batch.
     """
-    if object_valid_mask is not None:
-        pred_logits = pred_logits[object_valid_mask]
-        targets = targets[object_valid_mask]
-        sample_weight = sample_weight[object_valid_mask] if sample_weight is not None else None
+    batch_size, num_objects, num_constituents = targets.shape
 
-    loss = F.binary_cross_entropy_with_logits(pred_logits, targets, weight=sample_weight, reduction="none")
+    if object_valid_mask is None:
+        object_valid_mask = torch.ones(batch_size, num_objects, dtype=torch.bool, device=targets.device)
+    if input_pad_mask is None:
+        input_pad_mask = torch.ones(batch_size, num_constituents, dtype=targets.dtype, device=targets.device)
 
-    # Apply input padding mask
-    if input_pad_mask is not None:
-        loss = loss * input_pad_mask.unsqueeze(1)
+    loss = F.binary_cross_entropy_with_logits(pred_logits, targets, reduction="none")
+    if sample_weight is not None:
+        loss = loss * sample_weight
 
-    # normalise by valid elements such that each mask contributes equally
-    if input_pad_mask is not None:
-        valid_counts = input_pad_mask.sum(-1, keepdim=True)
-        loss = loss.sum(-1) / valid_counts
-        return loss.mean()
-    return loss.mean(-1).mean()
+    obj_mask = object_valid_mask.unsqueeze(-1).to(loss.dtype)
+    inp_mask = input_pad_mask.unsqueeze(1).to(loss.dtype)
+    loss = loss * obj_mask * inp_mask
+
+    valid_inputs = inp_mask.sum(-1).clamp_min(1.0)
+    per_obj = loss.sum(-1) / valid_inputs
+
+    valid_objs = object_valid_mask.sum(-1).clamp_min(1)
+    per_batch = (per_obj * object_valid_mask).sum(-1) / valid_objs
+
+    return per_batch.mean()
```

Comment:
is this change just to avoid nans for zero constituents, and otherwise leaves the loss unchanged? since I think multiplying by the `object_valid_mask` is less performant than indexing by it early, maybe we can just check the zero constituent case manually?

Having said that this impl is definitely cleaner, so if the perf impact is not large we could go with this. I know focal at least is pretty heavy on memory though,,